zerver/lib/actions.py

```diff
@@ -897,7 +898,9 @@ def do_set_realm_property(
         UserProfile.objects.bulk_update(user_profiles, ["email"])
 
         for user_profile in user_profiles:
-            flush_user_profile(sender=UserProfile, instance=user_profile)
+            transaction.on_commit(
+                lambda: flush_user_profile(sender=UserProfile, instance=user_profile)
```

Comment:
Should this be in on `on_commit`? I think so, but worth double-checking

Reply:
I think so.  I think we possibly want to do something like this: 

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index e3f39d9a18..c8e6a0ad1b 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -893,14 +893,14 @@ def do_set_realm_property(
         user_profiles = UserProfile.objects.filter(realm=realm, is_bot=False)
         for user_profile in user_profiles:
             user_profile.email = get_display_email_address(user_profile)
-            # TODO: Design a bulk event for this or force-reload all clients
-            send_user_email_update_event(user_profile)
         UserProfile.objects.bulk_update(user_profiles, ["email"])
 
         for user_profile in user_profiles:
             transaction.on_commit(
                 lambda: flush_user_profile(sender=UserProfile, instance=user_profile)
             )
+            # TODO: Design a bulk event for this or force-reload all clients
+            send_user_email_update_event(user_profile)
 
 
 def do_set_realm_authentication_methods(
```

Because the sequence we want is: (1) transaction commits, (2) flush all caches, (3) send update events.  That way, we don't have to consider possible bugs wherein the events system ends up interacting with a pre-cache-flush version of the update.  I'll apply this change since it's quick.

Of course, we aren't achieving that goal in any case, because the cache flush of the realm that happens on the `realm.save` is happening immediately.  We should probably serialize our understanding of wanting our on-save hooks to run with `transaction.on_commit` problem to an issue labeled with "difficult", while we still sorta remember it.